README.md
# Honeybadger's Webpack Source Map Plugin

[![Build Status](https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fhoneybadger-io%2Fhoneybadger-webpack%2Fbadge%3Fref%3Dmaster&style=flat)](https://actions-badge.atrox.dev/honeybadger-io/honeybadger-webpack/goto?ref=master)
[![npm version](https://badge.fury.io/js/%40honeybadger-io%2Fwebpack.svg)](https://badge.fury.io/js/%40honeybadger-io%2Fwebpack)

[Webpack](https://webpack.js.org/) plugin to upload JavaScript
sourcemaps to [Honeybadger](https://docs.honeybadger.io/guides/source-maps.html). You can also send [deployment notifications](https://docs.honeybadger.io/api/deployments.html).

Word Up! to the [thredUP](https://github.com/thredup) development team for a
similar webpack plugin they have authored.

## Installation

```
npm install @honeybadger-io/webpack --save-dev
```

## Configuration

### Plugin parameters

These plugin parameters correspond to the Honeybadger [Source Map Upload API](https://docs.honeybadger.io/guides/source-maps.html) and [Deployments API](https://docs.honeybadger.io/api/deployments.html).

<dl>
  <dt><code>apiKey</code> (required)</dt>
  <dd>The API key of your Honeybadger project</dd>

  <dt><code>assetsUrl</code> (required)</dt>
  <dd>The base URL to production assets (scheme://host/path)<code>*</code><a href="https://docs.honeybadger.io/guides/source-maps.html#wildcards">wildcards</a> are supported. The plugin combines <code>assetsUrl</code> with the generated minified js file name to build the API parameter <code>minified_url</code></dd>

  <dt><code>endpoint</code> (optional &mdash; default: "https://api.honeybadger.io/v1/source_maps")</dt>
  <dd>Where to upload your sourcemaps to. Perhaps you have a self hosted
  sourcemap server you would like to upload your sourcemaps to instead
  of honeybadger.</dd>

  <dt><code>revision</code> (optional &mdash; default: "master")</dt>
  <dd>The deploy revision (i.e. commit hash) that your source map applies to. This could also be a code version. For best results, set it to something unique every time your code changes. <a href="https://docs.honeybadger.io/lib/javascript/guides/using-source-maps.html#versioning-your-project">See the Honeybadger docs for examples.</a></dd>

  <dt><code>silent</code> (optional &mdash; default: "null/false")</dt>
  <dd>If true, silence log information emitted by the plugin.</dd>

  <dt><code>ignoreErrors</code> (optional &mdash; default: false)</dt>
  <dd>If true, webpack compilation errors are treated as warnings.</dd>

  <dt><code>retries</code> (optional &mdash; default: 3, max: 10)</dt>
  <dd>This package implements fetch retry functionality via
  https://github.com/vercel/fetch-retry </br>
  Retrying helps fix issues like `ECONNRESET` and `SOCKETTIMEOUT`
  errors and retries on 429 and 500 errors as well.
  </dd>

  <dt><code>workerCount</code> (optional &mdash; default: 5, min: 1)</dt>
  <dd>Sourcemaps are uploaded in parallel by a configurable number of 
  workers. Increase or decrease this value to configure how many sourcemaps
  are being uploaded in parallel.</br>
  Limited parallelism helps with connection issues in Docker environments.</dd>

  <dt><code>deploy</code> (optional &mdash; default: false)</dt>
  <dd>
  Configuration for deployment notifications. To disable deployment notifications, ignore this option. To enable deployment notifications, set this to <code>true</code>, or to an object containing any of these fields (see the <a href="https://docs.honeybadger.io/api/deployments.html">API reference</a>):</br>

  <dl>
    <dt><code>environment</code></dt>
    <dd>The environment name, for example, "production"</dd>
    <dt><code>repository</code></dt>
    <dd>The base URL of the VCS repository (HTTPS-style), for example, "https://github.com/yourusername/yourrepo"</dd>
    <dt><code>localUsername</code></dt>
    <dd>The name of the user that triggered this deploy, for example, "Jane"</dd>
  </dl>
  </dd>
</dl>

### Vanilla webpack.config.js

```javascript
const HoneybadgerSourceMapPlugin = require('@honeybadger-io/webpack')
const ASSETS_URL = 'https://cdn.example.com/assets';
const webpackConfig = {
  plugins: [new HoneybadgerSourceMapPlugin({
    apiKey: 'abc123',
    assetsUrl: ASSETS_URL,
    revision: 'master',
    // You can also enable deployment notifications:
    deploy: {
       environment: process.env.NODE_ENV,
       repository: "https://github.com/yourusername/yourrepo"
    }
  })]
}
```

### Rails Webpacker config/webpack/environment.js

```javascript
const { environment } = require('@rails/webpacker')
const HoneybadgerSourceMapPlugin = require('@honeybadger-io/webpack')

// Assumes Heroku / 12-factor application style ENV variables
// named GIT_COMMIT, HONEYBADGER_API_KEY, ASSETS_URL
const revision = process.env.GIT_COMMIT || 'master'

environment.plugins.append(
  'HoneybadgerSourceMap',
  new HoneybadgerSourceMapPlugin({
    apiKey: process.env.HONEYBADGER_API_KEY,
    assetsUrl: process.env.ASSETS_URL,
    silent: false,
    ignoreErrors: false,
    revision: revision
  }))

module.exports = environment
```

## Changelog

See https://github.com/honeybadger-io/honeybadger-webpack/blob/master/CHANGELOG.md

## Contributing

1. Fork it.
2. Create a topic branch `git checkout -b my_branch`
3. Commit your changes `git commit -am "Boom"`
3. Push to your branch `git push origin my_branch`
4. Send a [pull request](https://github.com/honeybadger-io/honeybadger-webpack/pulls)

## Development

1. Run `npm install`
2. Run the tests with `npm test`
3. Build/test on save with `npm run build:watch` and `npm run test:watch`

## Releasing

1. With a clean working tree, use `npm version [new version]` to bump the version,
   commit the changes, tag the release, and push to GitHub. See `npm help version`
   for documentation.
2. To publish the release, use `npm publish`. See `npm help publish` for
   documentation.

## License

The Honeybadger's Webpack Source Map Plugin is MIT licensed. See the
[MIT-LICENSE](https://raw.github.com/honeybadger-io/honeybadger-webpack/master/MIT-LICENSE)
file in this repository for details.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [1.5.1] - 2021-11-19
### Fixed
- Don't upload source maps if app is running with webpack-dev-server (#325)

## [1.5.0] - 2021-06-21
### Added
- Support for sending deployment notifications

## [1.4.0] - 2021-04-20
### Added
- Add worker support when uploading sourcemaps. Defaults to 5 files
  being uploaded in parallel.

## [1.3.0] - 2021-04-13
### Added
- Add retry functionality for fetch requests via
  [fetch-retry](https://github.com/vercel/fetch-retry)
- Add a retry option that defaults to 3, with a max number of retries
  of 10.
- Add a warning if no assets will be uploaded. Uses console.info instead
  of process.stdout.write.
- Add a configurable `endpoint` to the constructor, defaults to
  `https://api.honeybadger.io/v1/source_maps`
- Add a check for auxiliary files for Webpack 5 compatibility
- Add Webpack 5 compatibility
- Make Webpack 4+ a peerDependency

### Fixed
- fetch separates response errors from network errors.
  400+ status codes are treated separately from actual network errors.
- Attempt to reduce `ECONNRESET` and `SOCKETTIMEOUT` errors by
  using `fetch-retry`

## [1.2.0] - 2019-12-18
### Changed
- [Requires Webpack 4.39] Use assetEmitted hook to mitigate `futureEmitAssets = true` -@qnighy (#122)

### Fixed
- Dependency & security updates


test/helpers.test.js
/* eslint-env mocha */

import chai from 'chai'
import { REQUIRED_FIELDS } from '../src/constants'
import * as helpers from '../src/helpers'

const expect = chai.expect

describe('helpers', function () {
  describe('handleError', function () {
    it('should return an array of length 1 given a single error', function () {
      const result = helpers.handleError(new Error('required field missing'))
      expect(result).to.be.an.instanceof(Array)
      expect(result.length).to.eq(1)
    })

    it('should return an array of length 2 given an array of length 2', function () {
      const result = helpers.handleError([
        new Error('required field missing'),
        new Error('request failed')
      ])
      expect(result).to.an.instanceof(Array)
      expect(result.length).to.eq(2)
    })

    it('should prefix message of single error', function () {
      const result = helpers.handleError(new Error('required field missing'), 'Plugin')
      expect(result.length).to.eq(1)
      expect(result[0]).to.include({ message: 'Plugin: required field missing' })
    })

    it('should prefix message of an array of errors', function () {
      const result = helpers.handleError([
        new Error('required field missing'),
        new Error('request failed')
      ], 'Plugin')
      expect(result.length).to.eq(2)
      expect(result[0]).to.include({ message: 'Plugin: required field missing' })
    })

    it('should default prefix to "HoneybadgerSourceMapPlugin"', function () {
      const result = helpers.handleError(new Error('required field missing'))
      expect(result.length).to.eq(1)
      expect(result[0]).to.include({
        message: 'HoneybadgerSourceMapPlugin: required field missing'
      })
    })

    it('should handle null', function () {
      const result = helpers.handleError(null)
      expect(result).to.be.an('array')
      expect(result).to.be.empty
    })

    it('should handle empty []', function () {
      const result = helpers.handleError([])
      expect(result).to.be.an('array')
      expect(result).to.be.empty
    })
  })

  describe('validateOptions', function () {
    it('should return null if all required options are supplied', function () {
      const options = {
        apiKey: 'abcd1234',
        revision: 'fab5a8727c70647dcc539318b5b3e9b0cb8ae17b',
        assetsUrl: 'https://cdn.example.com/assets'
      }
      const result = helpers.validateOptions(options)
      expect(result).to.eq(null)
    })

    it('should return an error if apiKey is not supplied', function () {
      const options = {
        revision: 'fab5a8727c70647dcc539318b5b3e9b0cb8ae17b',
        assetsUrl: 'https://cdn.example.com/assets'
      }
      const result = helpers.validateOptions(options)
      expect(result).to.be.an.instanceof(Array)
      expect(result.length).to.eq(1)
      expect(result[0]).to.be.an.instanceof(Error)
      expect(result[0]).to.include({ message: 'required field, \'apiKey\', is missing.' })
    })

    it('should return an error if assetsUrl is not supplied', function () {
      const options = {
        apiKey: 'abcd1234',
        revision: 'fab5a8727c70647dcc539318b5b3e9b0cb8ae17b'
      }
      const result = helpers.validateOptions(options)
      expect(result).to.be.an.instanceof(Array)
      expect(result.length).to.eq(1)
      expect(result[0]).to.be.an.instanceof(Error)
      expect(result[0]).to.include({ message: 'required field, \'assetsUrl\', is missing.' })
    })

    it('should handle multiple missing required options', function () {
      const options = {}
      const result = helpers.validateOptions(options)
      expect(result).to.be.an.instanceof(Array)
      expect(result.length).to.eq(REQUIRED_FIELDS.length)
    })

    it('should handle null for options', function () {
      const result = helpers.validateOptions(null)
      expect(result).to.be.an.instanceof(Array)
      expect(result.length).to.eq(REQUIRED_FIELDS.length)
    })

    it('should handle no options passed', function () {
      const result = helpers.validateOptions()
      expect(result).to.be.an.instanceof(Array)
      expect(result.length).to.eq(REQUIRED_FIELDS.length)
    })
  })
})


test/integration.test.js
/* eslint-env mocha */

import HoneybadgerSourceMapPlugin from '../src/HoneybadgerSourceMapPlugin'
import webpack from 'webpack'
import chai from 'chai'
import path from 'path'
import nock from 'nock'
import sinon from 'sinon'

const expect = chai.expect

const consoleInfo = sinon.stub(console, 'info')
const ASSETS_URL = 'https://cdn.example.com/assets'
const TEST_ENDPOINT = 'https://api.honeybadger.io'
const SOURCEMAP_PATH = '/v1/source_maps'
const DEPLOY_PATH = '/v1/deploys'

it('only uploads source maps if no deploy config', function (done) {
  nock(TEST_ENDPOINT)
    .post(SOURCEMAP_PATH)
    .reply(201, JSON.stringify({ status: 'OK' }))

  const webpackConfig = {
    mode: 'development',
    entry: path.join(__dirname, 'fixtures/uncompiled.js'),
    output: {
      path: path.join(__dirname, '../tmp/')
    },
    devtool: 'source-map',
    plugins: [new HoneybadgerSourceMapPlugin({
      apiKey: 'abc123',
      retries: 0,
      assetsUrl: ASSETS_URL,
      revision: 'master'
    })]
  }
  webpack(webpackConfig, (err, stats) => {
    expect(err).to.eq(null)

    const info = stats.toJson('errors-warnings')
    expect(info.errors.length).to.equal(0)
    expect(info.warnings.length).to.equal(0)

    expect(consoleInfo.calledWith('Uploaded main.js.map to Honeybadger API')).to.eq(true)
    expect(consoleInfo.calledWith('Successfully sent deploy notification to Honeybadger API.')).to.eq(false)
    done()
  })
})

it('uploads source maps and sends deployment notification if configured', function (done) {
  nock(TEST_ENDPOINT)
    .post(SOURCEMAP_PATH)
    .reply(201, JSON.stringify({ status: 'OK' }))
  nock(TEST_ENDPOINT)
    .post(DEPLOY_PATH)
    .reply(201, JSON.stringify({ status: 'OK' }))

  const webpackConfig = {
    mode: 'development',
    entry: path.join(__dirname, 'fixtures/uncompiled.js'),
    output: {
      path: path.join(__dirname, '../tmp/')
    },
    devtool: 'source-map',
    plugins: [new HoneybadgerSourceMapPlugin({
      apiKey: 'abc123',
      retries: 0,
      assetsUrl: ASSETS_URL,
      revision: 'master',
      deploy: {
        environment: 'production',
        repository: 'https://cdn.example.com',
        localUsername: 'Jane'
      }
    })]
  }

  webpack(webpackConfig, (err, stats) => {
    expect(err).to.eq(null)

    const info = stats.toJson('errors-warnings')
    expect(info.errors.length).to.equal(0)
    expect(info.warnings.length).to.equal(0)

    expect(consoleInfo.calledWith('Uploaded main.js.map to Honeybadger API')).to.eq(true)
    expect(consoleInfo.calledWith('Successfully sent deploy notification to Honeybadger API.')).to.eq(true)
    done()
  })
})


test/fixtures/uncompiled.js

import find from 'lodash.find'


ship.config.js
/* eslint-disable @typescript-eslint/no-var-requires */
const fs = require('fs');
const semver = require('semver')
const keepachangelog = require('keepachangelog')

// eslint-disable-next-line no-undef
module.exports = {
  updateChangelog: false,
  formatCommitMessage: ({ version, _releaseType, _baseBranch }) => `Release v${version}`,
  formatPullRequestTitle: ({ version, _releaseType }) => `Release v${version}`,
  getNextVersion: ({ _revisionRange, _commitTitles, _commitBodies, currentVersion, dir }) => {
    const changelogFile = `${dir}/CHANGELOG.md`
    const content = fs.readFileSync(changelogFile, 'utf8')
    const changelog = keepachangelog.parse(content)
    return getNextVersion(changelog, currentVersion)
  },
  beforeCommitChanges: ({ nextVersion, _releaseType, _exec, dir }) => {
    const changelogFile = `${dir}/CHANGELOG.md`
    const content = fs.readFileSync(changelogFile, 'utf8')
    const changelog = keepachangelog.parse(content)
    changelog.addRelease(nextVersion)
    fs.writeFileSync(changelogFile, changelog.build(), 'utf8')
  },
  shouldPrepare: ({ nextVersion, commitNumbersPerType }) => {
    return !!nextVersion
  }
}


function getNextVersion(changelog, currentVersion, releaseTag = 'latest') {
  const parsedVersion = semver.parse(currentVersion)
  const upcomingRelease = changelog.getRelease('upcoming') || { version: 'upcoming' }

  let releaseType
  if (upcomingRelease.Changed?.length > 0) {
    releaseType = 'major'
  } else if (upcomingRelease.Added?.length > 0) {
    releaseType = 'minor'
  } else if (upcomingRelease.Fixed?.length > 0) {
    releaseType = 'patch'
  } else {
    return null
  }

  if (releaseTag !== 'latest') {
    if (parsedVersion.prerelease.length) {
      parsedVersion.inc('prerelease', releaseTag)
    } else {
      parsedVersion.inc(releaseType)
      parsedVersion.prerelease = [ releaseTag, 0 ]
      parsedVersion.format()
    }
  } else {
    parsedVersion.inc(releaseType)
  }

  return parsedVersion.version
}


src/HoneybadgerSourceMapPlugin.js
import { promises as fs } from 'fs'
import { join } from 'path'
import nodeFetch from 'node-fetch'
import fetchRetry from '@vercel/fetch-retry'
import VError from 'verror'
import find from 'lodash.find'
import reduce from 'lodash.reduce'
import FormData from 'form-data'
import { handleError, validateOptions } from './helpers'
import { ENDPOINT, DEPLOY_ENDPOINT, PLUGIN_NAME, MAX_RETRIES, MIN_WORKER_COUNT } from './constants'
import { resolvePromiseWithWorkers } from './resolvePromiseWithWorkers'

const fetch = fetchRetry(nodeFetch)

/**
 * @typedef {Object} DeployObject
 * @property {?string} environment - production, development, staging, etc
 * @property {?string} repository - URL for repository IE: https://github.com/foo/bar
 * @property {?string} localUsername - The name of the user deploying. IE: Jane
 */

class HoneybadgerSourceMapPlugin {
  constructor ({
    apiKey,
    assetsUrl,
    endpoint = ENDPOINT,
    revision = 'master',
    silent = false,
    ignoreErrors = false,
    retries = 3,
    workerCount = 5,
    deploy = false
  }) {
    this.apiKey = apiKey
    this.assetsUrl = assetsUrl
    this.endpoint = endpoint
    this.revision = revision
    this.silent = silent
    this.ignoreErrors = ignoreErrors
    this.workerCount = Math.max(workerCount, MIN_WORKER_COUNT)
    /** @type DeployObject|boolean */
    this.deploy = deploy
    this.retries = retries

    if (this.retries > MAX_RETRIES) {
      this.retries = MAX_RETRIES
    }
  }

  async afterEmit (compilation) {
    if (this.isDevServerRunning()) {
      if (!this.silent) {
        console.info('\nHoneybadgerSourceMapPlugin will not upload source maps because webpack-dev-server is running.')
      }

      return
    }

    const errors = validateOptions(this)

    if (errors) {
      compilation.errors.push(...handleError(errors))
      return
    }

    try {
      await this.uploadSourceMaps(compilation)
      await this.sendDeployNotification()
    } catch (err) {
      if (!this.ignoreErrors) {
        compilation.errors.push(...handleError(err))
      } else if (!this.silent) {
        compilation.warnings.push(...handleError(err))
      }
    }
  }

  isDevServerRunning () {
    return process.env.WEBPACK_DEV_SERVER === 'true'
  }

  apply (compiler) {
    compiler.hooks.afterEmit.tapPromise(PLUGIN_NAME, this.afterEmit.bind(this))
  }

  // eslint-disable-next-line class-methods-use-this
  getAssetPath (compilation, name) {
    return join(
      compilation.getPath(compilation.compiler.outputPath),
      name.split('?')[0]
    )
  }

  getSource (compilation, name) {
    const path = this.getAssetPath(compilation, name)
    return fs.readFile(path, { encoding: 'utf-8' })
  }

  getAssets (compilation) {
    const { chunks } = compilation.getStats().toJson()

    return reduce(chunks, (result, chunk) => {
      const sourceFile = find(chunk.files, file => /\.js$/.test(file))

      // Webpack 4 using chunk.files, Webpack 5 uses chunk.auxiliaryFiles
      // https://webpack.js.org/blog/2020-10-10-webpack-5-release/#stats
      const sourceMap = (chunk.auxiliaryFiles || chunk.files).find(file =>
        /\.js\.map$/.test(file)
      )

      if (!sourceFile || !sourceMap) {
        return result
      }

      return [
        ...result,
        { sourceFile, sourceMap }
      ]
    }, [])
  }

  getUrlToAsset (sourceFile) {
    if (typeof sourceFile === 'string') {
      const sep = this.assetsUrl.endsWith('/') ? '' : '/'
      return `${this.assetsUrl}${sep}${sourceFile}`
    }
    return this.assetsUrl(sourceFile)
  }

  async uploadSourceMap (compilation, { sourceFile, sourceMap }) {
    const errorMessage = `failed to upload ${sourceMap} to Honeybadger API`

    let sourceMapSource
    let sourceFileSource

    try {
      sourceMapSource = await this.getSource(compilation, sourceMap)
      sourceFileSource = await this.getSource(compilation, sourceFile)
    } catch (err) {
      throw new VError(err, err.message)
    }

    const form = new FormData()
    form.append('api_key', this.apiKey)
    form.append('minified_url', this.getUrlToAsset(sourceFile))
    form.append('minified_file', sourceFileSource, {
      filename: sourceFile,
      contentType: 'application/javascript'
    })
    form.append('source_map', sourceMapSource, {
      filename: sourceMap,
      contentType: 'application/octet-stream'
    })
    form.append('revision', this.revision)

    let res
    try {
      res = await fetch(this.endpoint, {
        method: 'POST',
        body: form,
        redirect: 'follow',
        opts: {
          retries: this.retries,
          // Max timeout between retries, in milliseconds
          maxTimeout: 1000
        }
      })
    } catch (err) {
      // network / operational errors. Does not include 404 / 500 errors
      throw new VError(err, errorMessage)
    }

    // >= 400 responses
    if (!res.ok) {
      // Attempt to parse error details from response
      let details
      try {
        const body = await res.json()

        if (body && body.error) {
          details = body.error
        } else {
          details = `${res.status} - ${res.statusText}`
        }
      } catch (parseErr) {
        details = `${res.status} - ${res.statusText}`
      }

      throw new Error(`${errorMessage}: ${details}`)
    }

    // Success
    if (!this.silent) {
      // eslint-disable-next-line no-console
      console.info(`Uploaded ${sourceMap} to Honeybadger API`)
    }
  }

  uploadSourceMaps (compilation) {
    const assets = this.getAssets(compilation)

    if (assets.length <= 0) {
      // We should probably tell people they're not uploading assets.
      // this is also an open issue on Rollbar sourcemap plugin
      // https://github.com/thredup/rollbar-sourcemap-webpack-plugin/issues/39
      if (!this.silent) {
        console.info(this.noAssetsFoundMessage)
      }

      return
    }

    console.info('\n')

    // On large projects source maps should not all be uploaded at the same time,
    // but in parallel with a reasonable worker count in order to avoid network issues
    return resolvePromiseWithWorkers(
      assets.map(asset => () => this.uploadSourceMap(compilation, asset)),
      this.workerCount
    )
  }

  async sendDeployNotification () {
    if (this.deploy === false || this.apiKey == null) {
      return
    }

    let body

    if (this.deploy === true) {
      body = {
        deploy: {
          revision: this.revision
        }
      }
    } else if (typeof this.deploy === 'object' && this.deploy !== null) {
      body = {
        deploy: {
          revision: this.revision,
          repository: this.deploy.repository,
          local_username: this.deploy.localUsername,
          environment: this.deploy.environment
        }
      }
    }

    const errorMessage = 'Unable to send deploy notification to Honeybadger API.'
    let res

    try {
      res = await fetch(DEPLOY_ENDPOINT, {
        method: 'POST',
        headers: {
          'X-API-KEY': this.apiKey,
          'Content-Type': 'application/json',
          Accept: 'application/json'
        },
        body: JSON.stringify(body),
        redirect: 'follow',
        opts: {
          retries: this.retries,
          // Max timeout between retries, in milliseconds
          maxTimeout: 1000
        }
      })
    } catch (err) {
      // network / operational errors. Does not include 404 / 500 errors
      if (!this.ignoreErrors) {
        throw new VError(err, errorMessage)
      }
    }

    // >= 400 responses
    if (!res.ok) {
      // Attempt to parse error details from response
      let details
      try {
        const body = await res.json()

        if (body && body.error) {
          details = body.error
        } else {
          details = `${res.status} - ${res.statusText}`
        }
      } catch (parseErr) {
        details = `${res.status} - ${res.statusText}`
      }

      if (!this.ignoreErrors) {
        throw new Error(`${errorMessage}: ${details}`)
      }
    }

    if (!this.silent) {
      console.info('Successfully sent deploy notification to Honeybadger API.')
    }
  }

  get noAssetsFoundMessage () {
    return '\nHoneybadger could not find any sourcemaps. Nothing will be uploaded.'
  }
}

module.exports = HoneybadgerSourceMapPlugin
